Success response headers are not added to the response

**Describe the bug**
Since kuadrant-operator-catalog nightly builds from `28-29.01.25` our tests that should return custom success response headers started failing. They don't add these headers into the response anymore. Here is the [most simple test](https://github.com/Kuadrant/testsuite/blob/4b95ca14fe0ee697f8f0e615adbb0be21a49071d/testsuite/tests/singlecluster/authorino/response/test_simple_response.py#L19) for responses we got that is failing with `AssertionError: Header from response (Header) is missing; assert None is not None`

We were experiencing the same failures a few months ago, which ended up being a bug inside the wasm-shim, and was fixed by [this](https://github.com/Kuadrant/wasm-shim/pull/143) PR.

**Help us Reproduce it**

AuthPolicy:
```yaml
spec:
  rules:
    authentication:
      default:
        credentials:
          authorizationHeader:
            prefix: Bearer
        jwt:
          issuerUrl: 'http://1.2.3.4:8080/realms/realm-averevki--eyt7'
          ttl: 0
        metrics: false
        priority: 0
    response:
      success:
        headers:
          header:
            json:
              properties:
                anything:
                  value: one
            metrics: false
            priority: 0
  targetRef:
    group: gateway.networking.k8s.io
    kind: HTTPRoute
    name: route-averevki--obnu
```
<details>
<summary>Gateway</summary>

```yaml
spec:
  gatewayClassName: istio
  listeners:
    - allowedRoutes:
        namespaces:
          from: Same
      hostname: '*.apps.kua.redhat.com'
      name: api
      port: 80
      protocol: HTTP
```
</details>

<details>
<summary>HTTPRoute</summary>

```yaml
spec:
  hostnames:
    - hostname-averevki--nbep-kuadrant.apps.kua.redhat.com
  parentRefs:
    - group: gateway.networking.k8s.io
      kind: Gateway
      name: gw-averevki--ljiu
  rules:
    - backendRefs:
        - group: ''
          kind: Service
          name: httpbin-averevki--ihex
          namespace: kuadrant
          port: 8080
          weight: 1
      matches:
        - path:
            type: PathPrefix
            value: /
```
</details>

**Expected behavior**

Return headers I specify inside AuthPolicy with a successful response

**Environment (please complete the following information):**

kuadrant-operator image: `nightly-04-02-2025`
authorino-operator image: `e1dc399c6bcb2fb6d28b2f3e94ce49bb5b63e06c`
authorino image: `b2bde91bfc64b1d140f740069ba36ce81a3f3349`
wasm-shim image: `751af05ea5e6c440d6254480d669bc5861f457eb`

--- a/src/auth_action.rs
+++ b/src/auth_action.rs
@@ -1,7 +1,7 @@
 use crate::configuration::{Action, FailureMode, Service};
 use crate::data::{store_metadata, Predicate, PredicateVec};
 use crate::envoy::{CheckResponse, CheckResponse_oneof_http_response, HeaderValueOption};
-use crate::service::{GrpcErrResponse, GrpcService, Headers};
+use crate::service::{GrpcErrResponse, GrpcService, HeaderKind, Headers};
 use log::debug;
 use std::rc::Rc;
 
@@ -45,7 +45,7 @@ impl AuthAction {
     pub fn process_response(
         &self,
         check_response: CheckResponse,
-    ) -> Result<Headers, GrpcErrResponse> {
+    ) -> Result<HeaderKind, GrpcErrResponse> {
         //todo(adam-cattermole):hostvar resolver?
         // store dynamic metadata in filter state
         debug!("process_response(auth): store_metadata");
@@ -58,7 +58,7 @@ impl AuthAction {
                     FailureMode::Deny => Err(GrpcErrResponse::new_internal_server_error()),
                     FailureMode::Allow => {
                         debug!("process_response(auth): continuing as FailureMode Allow");
-                        Ok(Vec::default())
+                        Ok(HeaderKind::Request(Vec::default()))
                     }
                 }
             }
@@ -87,7 +87,9 @@ impl AuthAction {
                 if !ok_response.get_query_parameters_to_remove().is_empty() {
                     panic!("process_response(auth): response contained query_parameters_to_remove which is unsupported!")
                 }
-                Ok(Self::get_header_vec(ok_response.get_headers()))
+                Ok(HeaderKind::Request(Self::get_header_vec(
+                    ok_response.get_headers(),
+                )))
             }
         }
     }
@@ -228,21 +230,30 @@ mod test {
         let result = auth_action.process_response(ok_response_without_headers);
         assert!(result.is_ok());
 
-        let headers = result.expect("is ok");
-        assert!(headers.is_empty());
+        match result.expect("is ok") {
+            HeaderKind::Response(_headers) => {
+                panic!("check_response should not return Response headers")
+            }
+            HeaderKind::Request(headers) => assert!(headers.is_empty()),
+        }
 
         let ok_response_with_header =
             build_check_response(StatusCode::OK, Some(vec![("my_header", "my_value")]), None);
         let result = auth_action.process_response(ok_response_with_header);
         assert!(result.is_ok());
 
-        let headers = result.expect("is ok");
-        assert!(!headers.is_empty());
-
-        assert_eq!(
-            headers[0],
-            ("my_header".to_string(), "my_value".to_string())
-        );
+        match result.expect("is ok") {
+            HeaderKind::Response(_headers) => {
+                panic!("check_response should not return Response headers")
+            }
+            HeaderKind::Request(headers) => {
+                assert!(!headers.is_empty());
+                assert_eq!(
+                    headers[0],
+                    ("my_header".to_string(), "my_value".to_string())
+                );
+            }
+        }
     }
 
     #[test]

--- a/src/filter/kuadrant_filter.rs
+++ b/src/filter/kuadrant_filter.rs
@@ -1,9 +1,9 @@
 use crate::action_set_index::ActionSetIndex;
 use crate::filter::operations::{
-    GrpcMessageReceiverOperation, GrpcMessageSenderOperation, HeadersOperation, Operation,
+    GrpcMessageReceiverOperation, GrpcMessageSenderOperation, Operation,
 };
 use crate::runtime_action_set::RuntimeActionSet;
-use crate::service::{GrpcErrResponse, GrpcRequest, HeaderResolver};
+use crate::service::{GrpcErrResponse, GrpcRequest, HeaderResolver, Headers};
 use log::{debug, warn};
 use proxy_wasm::traits::{Context, HttpContext};
 use proxy_wasm::types::{Action, Status};
@@ -16,7 +16,8 @@ pub(crate) struct KuadrantFilter {
     header_resolver: Rc<HeaderResolver>,
 
     grpc_message_receiver_operation: Option<GrpcMessageReceiverOperation>,
-    headers_operations: Vec<HeadersOperation>,
+    response_headers_to_add: Option<Headers>,
+    request_headers_to_add: Option<Headers>,
 }
 
 impl Context for KuadrantFilter {
@@ -51,6 +52,9 @@ impl HttpContext for KuadrantFilter {
         #[cfg(feature = "debug-host-behaviour")]
         crate::data::debug_all_well_known_attributes();
 
+        // default action if we find no action_set where conditions apply
+        let mut action = Action::Continue;
+
         if let Some(action_sets) = self
             .index
             .get_longest_match_action_sets(self.request_authority().as_ref())
@@ -63,18 +67,23 @@ impl HttpContext for KuadrantFilter {
                     "#{} action_set selected {}",
                     self.context_id, action_set.name
                 );
-                return self.start_flow(Rc::clone(action_set));
+                action = self.start_flow(Rc::clone(action_set))
             }
         }
-        Action::Continue
+
+        if action == Action::Continue {
+            // the request headers are currently always None, however this is one of two phases
+            // where headers should be added
+            self.add_request_headers()
+        }
+        action
     }
 
     fn on_http_response_headers(&mut self, _num_headers: usize, _end_of_stream: bool) -> Action {
         debug!("#{} on_http_response_headers", self.context_id);
-        let headers_operations = mem::take(&mut self.headers_operations);
-        for op in headers_operations {
-            for (header, value) in &op.headers() {
-                self.add_http_response_header(header, value)
+        if let Some(response_headers) = mem::take(&mut self.response_headers_to_add) {
+            for (header, value) in response_headers {
+                self.add_http_response_header(header.as_str(), value.as_str())
             }
         }
         Action::Continue
@@ -116,7 +125,22 @@ impl KuadrantFilter {
             }
             Operation::AddHeaders(header_op) => {
                 debug!("handle_operation: AddHeaders");
-                self.headers_operations.push(header_op);
+                match header_op.into_inner() {
+                    crate::service::HeaderKind::Request(headers) => {
+                        if let Some(existing_headers) = self.request_headers_to_add.as_mut() {
+                            existing_headers.extend(headers);
+                        } else {
+                            warn!("Trying to add request headers after phase has ended!")
+                        }
+                    }
+                    crate::service::HeaderKind::Response(headers) => {
+                        if let Some(existing_headers) = self.response_headers_to_add.as_mut() {
+                            existing_headers.extend(headers);
+                        } else {
+                            warn!("Trying to add response headers after phase has ended!")
+                        }
+                    }
+                }
                 Action::Continue
             }
             Operation::Die(die_op) => {
@@ -126,6 +150,7 @@ impl KuadrantFilter {
             }
             Operation::Done() => {
                 debug!("handle_operation: Done");
+                self.add_request_headers();
                 self.resume_http_request();
                 Action::Continue
             }
@@ -171,6 +196,14 @@ impl KuadrantFilter {
         )
     }
 
+    fn add_request_headers(&mut self) {
+        if let Some(request_headers) = mem::take(&mut self.request_headers_to_add) {
+            for (header, value) in request_headers {
+                self.add_http_request_header(header.as_str(), value.as_str())
+            }
+        }
+    }
+
     pub fn new(
         context_id: u32,
         index: Rc<ActionSetIndex>,
@@ -181,7 +214,8 @@ impl KuadrantFilter {
             index,
             header_resolver,
             grpc_message_receiver_operation: None,
-            headers_operations: Vec::default(),
+            response_headers_to_add: Some(Vec::default()),
+            request_headers_to_add: Some(Vec::default()),
         }
     }
 }

--- a/src/filter/operations.rs
+++ b/src/filter/operations.rs
@@ -1,13 +1,13 @@
 use crate::configuration::FailureMode;
 use crate::filter::operations::Operation::SendGrpcRequest;
 use crate::runtime_action_set::RuntimeActionSet;
-use crate::service::{GrpcErrResponse, GrpcRequest, Headers, IndexedGrpcRequest};
+use crate::service::{GrpcErrResponse, GrpcRequest, HeaderKind, IndexedGrpcRequest};
 use std::rc::Rc;
 
 pub enum Operation {
     SendGrpcRequest(GrpcMessageSenderOperation),
     AwaitGrpcResponse(GrpcMessageReceiverOperation),
-    AddHeaders(HeadersOperation),
+    AddHeaders(HeaderOperation),
     Die(GrpcErrResponse),
     // Done indicates that we have no more operations and can resume the http request flow
     Done(),
@@ -60,7 +60,7 @@ impl GrpcMessageReceiverOperation {
             Ok((next_msg, headers)) => {
                 let mut operations = Vec::new();
                 if !headers.is_empty() {
-                    operations.push(Operation::AddHeaders(HeadersOperation::new(headers)))
+                    operations.push(Operation::AddHeaders(HeaderOperation::new(headers)))
                 }
                 operations.push(match next_msg {
                     None => Operation::Done(),
@@ -92,16 +92,16 @@ impl GrpcMessageReceiverOperation {
     }
 }
 
-pub struct HeadersOperation {
-    headers: Headers,
+pub struct HeaderOperation {
+    headers: HeaderKind,
 }
 
-impl HeadersOperation {
-    pub fn new(headers: Headers) -> Self {
+impl HeaderOperation {
+    pub fn new(headers: HeaderKind) -> Self {
         Self { headers }
     }
 
-    pub fn headers(self) -> Headers {
+    pub fn into_inner(self) -> HeaderKind {
         self.headers
     }
 }

--- a/src/ratelimit_action.rs
+++ b/src/ratelimit_action.rs
@@ -5,7 +5,7 @@ use crate::envoy::{
     HeaderValue, RateLimitDescriptor, RateLimitDescriptor_Entry, RateLimitResponse,
     RateLimitResponse_Code, StatusCode,
 };
-use crate::service::{GrpcErrResponse, GrpcService, Headers};
+use crate::service::{GrpcErrResponse, GrpcService, HeaderKind, Headers};
 use cel_interpreter::Value;
 use log::{debug, error};
 use protobuf::RepeatedField;
@@ -167,7 +167,7 @@ impl RateLimitAction {
     pub fn process_response(
         &self,
         rate_limit_response: RateLimitResponse,
-    ) -> Result<Vec<(String, String)>, GrpcErrResponse> {
+    ) -> Result<HeaderKind, GrpcErrResponse> {
         match rate_limit_response {
             RateLimitResponse {
                 overall_code: RateLimitResponse_Code::UNKNOWN,
@@ -178,7 +178,7 @@ impl RateLimitAction {
                     FailureMode::Deny => Err(GrpcErrResponse::new_internal_server_error()),
                     FailureMode::Allow => {
                         debug!("process_response(rl): continuing as FailureMode Allow");
-                        Ok(Vec::default())
+                        Ok(HeaderKind::Response(Vec::default()))
                     }
                 }
             }
@@ -201,7 +201,9 @@ impl RateLimitAction {
                 ..
             } => {
                 debug!("process_response(rl): received OK response");
-                Ok(Self::get_header_vec(additional_headers))
+                Ok(HeaderKind::Response(Self::get_header_vec(
+                    additional_headers,
+                )))
             }
         }
     }
@@ -425,13 +427,18 @@ mod test {
         let result = rl_action.process_response(ok_response_with_header);
         assert!(result.is_ok());
 
-        let headers = result.expect("is ok");
-        assert!(!headers.is_empty());
-
-        assert_eq!(
-            headers[0],
-            ("my_header".to_string(), "my_value".to_string())
-        );
+        match result.expect("is ok") {
+            HeaderKind::Response(headers) => {
+                assert!(!headers.is_empty());
+                assert_eq!(
+                    headers[0],
+                    ("my_header".to_string(), "my_value".to_string())
+                );
+            }
+            HeaderKind::Request(_headers) => {
+                panic!("ratelimitresponse should not return Request headers")
+            }
+        }
     }
 
     #[test]

--- a/src/runtime_action.rs
+++ b/src/runtime_action.rs
@@ -3,7 +3,7 @@ use crate::configuration::{Action, FailureMode, Service, ServiceType};
 use crate::ratelimit_action::RateLimitAction;
 use crate::service::auth::AuthService;
 use crate::service::rate_limit::RateLimitService;
-use crate::service::{GrpcErrResponse, GrpcRequest, GrpcService, Headers};
+use crate::service::{GrpcErrResponse, GrpcRequest, GrpcService, HeaderKind};
 use log::debug;
 use protobuf::Message;
 use std::collections::HashMap;
@@ -67,7 +67,7 @@ impl RuntimeAction {
         }
     }
 
-    pub fn process_response(&self, msg: &[u8]) -> Result<Headers, GrpcErrResponse> {
+    pub fn process_response(&self, msg: &[u8]) -> Result<HeaderKind, GrpcErrResponse> {
         match self {
             Self::Auth(auth_action) => match Message::parse_from_bytes(msg) {
                 Ok(check_response) => auth_action.process_response(check_response),
@@ -77,7 +77,7 @@ impl RuntimeAction {
                         FailureMode::Deny => Err(GrpcErrResponse::new_internal_server_error()),
                         FailureMode::Allow => {
                             debug!("process_response(auth): continuing as FailureMode Allow");
-                            Ok(Vec::default())
+                            Ok(HeaderKind::Request(Vec::default()))
                         }
                     }
                 }
@@ -90,7 +90,7 @@ impl RuntimeAction {
                         FailureMode::Deny => Err(GrpcErrResponse::new_internal_server_error()),
                         FailureMode::Allow => {
                             debug!("process_response(rl): continuing as FailureMode Allow");
-                            Ok(Vec::default())
+                            Ok(HeaderKind::Response(Vec::default()))
                         }
                     }
                 }

--- a/src/runtime_action_set.rs
+++ b/src/runtime_action_set.rs
@@ -1,7 +1,7 @@
 use crate::configuration::{ActionSet, Service};
 use crate::data::{Predicate, PredicateVec};
 use crate::runtime_action::RuntimeAction;
-use crate::service::{GrpcErrResponse, Headers, IndexedGrpcRequest};
+use crate::service::{GrpcErrResponse, HeaderKind, IndexedGrpcRequest};
 use std::collections::HashMap;
 use std::rc::Rc;
 
@@ -80,7 +80,7 @@ impl RuntimeActionSet {
         &self,
         index: usize,
         msg: &[u8],
-    ) -> Result<(Option<IndexedGrpcRequest>, Headers), GrpcErrResponse> {
+    ) -> Result<(Option<IndexedGrpcRequest>, HeaderKind), GrpcErrResponse> {
         self.runtime_actions[index]
             .process_response(msg)
             .map(|headers| {

--- a/src/service.rs
+++ b/src/service.rs
@@ -132,6 +132,20 @@ impl GrpcRequest {
 
 pub type Headers = Vec<(String, String)>;
 #[derive(Debug)]
+pub enum HeaderKind {
+    Response(Headers),
+    Request(Headers),
+}
+
+impl HeaderKind {
+    pub fn is_empty(&self) -> bool {
+        match self {
+            HeaderKind::Response(headers) | HeaderKind::Request(headers) => headers.is_empty(),
+        }
+    }
+}
+
+#[derive(Debug)]
 pub struct GrpcErrResponse {
     status_code: u32,
     response_headers: Headers,

--- a/utils/deploy/authconfig.yaml
+++ b/utils/deploy/authconfig.yaml
@@ -31,6 +31,12 @@ spec:
                 value: 42
               "missing":
                 value: null
+      headers:
+        "my-header":
+          json:
+            properties:
+              "my-key":
+                value: true
 ---
 apiVersion: v1
 kind: Secret

